registry/virtual.go

```diff
@@ -1,50 +1,455 @@
+		return err
+	}
+	db := s.VersionDB()
+	_, err = db.Delete(context.Background(), overwritten.ID, overwritten.Version)
+	return err
+}
+
+func storeOverwrittenTarball(s base.VirtualSpace, tarball string) (hash string, length int64, err error) {
+	file, err := os.Open(tarball)
+	if err != nil {
+		return "", 0, err
+	}
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	stats, err := file.Stat()
+	if err != nil {
+		return "", 0, err
+	}
+	length = stats.Size()
+
+	hasher := sha256.New()
+	if _, err := io.Copy(hasher, file); err != nil {
+		return "", 0, err
+	}
+	h := hasher.Sum(nil)
+	hash = hex.EncodeToString(h)
+
+	prefix := base.Prefix(s.Name)
+	if _, err := file.Seek(0, io.SeekStart); err != nil {
+		return "", 0, err
+	}
+
+	if err = base.Storage.EnsureExists(prefix); err != nil {
+		return "", 0, err
+	}
+	if err = base.Storage.Create(prefix, hash, "application/gzip", file); err != nil {
+		return "", 0, err
+	}
+
+	return hash, length, nil
+}
+
+func deleteOverwrittenTarball(s base.VirtualSpace, version *Version) error {
+	h, ok := version.AttachmentReferences["tarball"]
+	if ok {
+		prefix := base.Prefix(s.Name)
+		if err := base.Storage.Remove(prefix, h); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getOriginalTarball(space *space.Space, version *Version) (io.Reader, error) {
+	url, err := url.Parse(version.URL)
+	if err != nil {
+		return nil, err
+	}
+	filename := filepath.Base(url.Path)
+
+	att, err := FindVersionAttachment(space, version, filename)
+	if err != nil {
+		return nil, err
+	}
 
+	return att.Content, nil
+}
+
+func generateOverwrittenTarball(version *Version, overwrite map[string]interface{}, input io.Reader) (file string, manifest map[string]interface{}, icon string, err error) {
+	var newManifest map[string]interface{}
+
+	var originManifest Manifest
+	if err := json.Unmarshal(version.Manifest, &originManifest); err != nil {
+		return "", nil, "", err
+	}
+
+	iconFilename := originManifest.Icon
+	manifestFilename := "manifest." + version.Type
+
+	iconChecksum, iconOverwritten := overwrite["icon"].(string)
+	var iconContent *bytes.Buffer
+	if iconOverwritten {
+		iconContent, _, err = base.GlobalAssetStore.Get(iconChecksum)
+		if err != nil {
+			return "", nil, "", err
+		}
+	}
+	name, nameOverwritten := overwrite["name"].(string)
+
+	inputGzip, err := gzip.NewReader(input)
+	if err != nil {
+		return "", nil, "", err
+	}
+	defer inputGzip.Close()
+	inputTar := tar.NewReader(inputGzip)
+
+	prefix := fmt.Sprintf("%s_%s_*.tar.gz", version.Slug, version.Version)
+	outputFile, err := ioutil.TempFile("", prefix)
+	if err != nil {
+		return "", nil, "", err
+	}
+	file = outputFile.Name()
+	// Tricky return case from here. Never `return nil, err`.
+	// File is already created on filesystem, so we need to return it in all case to be able to clean it on the caller.
+	// We can'm `defer os.Remove` here, because caller need the file for storage…
```

Comment:
I think we are doing something messy here. In general, I try to make the same function creating a file and cleaning it. Here, I think it should be the responsibility of the caller to create the file.

Reply:
Let's do that.